Prevent duplicate sites in signup

niranjan-uma-shankar opened the pull request:
#### Changes proposed in this Pull Request

* Detect if signup flow was loaded via back button, and in that case do not create a new site when the signup flow is complete.
* Check pau2Xa-1Io-p2#comment-6759 for an explanation on how this implementation works
* There's an edge case in which a user could end up seeing the launch flow UI version when going to wordpress.com/start to add a new site. But it's going to be an extremely rare case, and a few things have to align, so we can safely not bother about this issue:
    - Firstly, when the user is on checkout page from signup flow, they should navigate to a different part of Calypso on the _same tab_ without using the close button on checkout header, or empty cart action.
    - Next, on the _same tab_, they should navigate to wordpress.com/start by directly typing the URL in the address bar, or clicking a link that opens the page(on the same tab).
    - All of this should happen in < 1 day since a new site was created (signup destination cookie will expire in 1 day).
Important to note that this issue will not be a permanent problem - the user just needs to open wordpress.com/start in a new tab to see the regular signup flow again. Again, this is a super edge case and not something to worry about IMO.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Scenario 1** (Verify that going back from checkout page to signup, you can select a new domain/plan and complete purchase without creating a duplicate site).

1. Starting logged out of wpcom, go through the signup at wordpress.com/start and select a paid plan.
2. On the checkout page, hit browser back. Verify that you see the launch flow UI version:

<img width="1245" alt="Screenshot 2020-08-27 at 5 38 18 PM" src="https://user-images.githubusercontent.com/1269602/91440367-1eeaef00-e88c-11ea-97bd-1a48f85ab060.png">

3. Test the following two scenarios:
    (A) Selecting a free plan should take you to customer home and duplicate site should not be created. 
    (B) Select a custom domain and paid plan and complete payment. Verify that the new domain and plan got added to your site and that duplicate sites were not created.

**Scenario 2** (When on signup via browser back from checkout, page reload should not affect the UI)

1.  Starting logged out of wpcom, go through the signup at wordpress.com/start and select a paid plan.
2.  On the checkout page, hit browser back. Verify that you see the launch flow UI version (screen shot in Scenario 1)
3. Reload the page and verify that you still see the launch flow UI version.

**Scenario 3** (Add new site should work, and open the regular signup flow)

1.  Starting logged out of wpcom, go through the signup at wordpress.com/start and select a paid plan.
2.  On the checkout page, click the close button in the masterbar or empty all cart items.
3. Click "Add new site" from Calypso's sidebar, and verify that the domain step shows the regular signup flow UI(and not the launch flow UI)

**Scenario 4**
1. Try browser back from checkout in other flows besides `onboarding`, like in /start/premium, or the plan-domain swap flow*(for this you need to be assigned to the test variant, check https://github.com/Automattic/wp-calypso/pull/43675#pullrequestreview-440991033 for details on how to enter the variant).
2. In all these flows, you should be able to add purchase to the existing site without creating duplicates sites.

**Scenario 5** (When signup destination cookie is not cleared, closing tab and reopening a new tab should kill `sessionStorage` so /start should open "Add new site" flow instead of launch flow UI version)

1.  Starting logged out of wpcom, go through the signup at wordpress.com/start and select a paid plan.
2. On the checkout page, close tab and in a new tab, open wordpress.com/start
3. You should be taken to the "Add new site" flow to create a new site, with the regular signup UI.

**Scenario 6**

Verify that https://github.com/Automattic/wp-calypso/issues/45862 is fixed.

niranjan-uma-shankar committed:
detect page load via history

niranjan-uma-shankar committed:
Show launch flow version of domain UI when hitting back button from checkout

niranjan-uma-shankar committed:
Store state in sessionStorage

niranjan-uma-shankar committed:
disable eslint

niranjan-uma-shankar committed:
1. Send site slug for add domain to cart 2. rename preventDuplicates to isManageSiteFlow

niranjan-uma-shankar committed:
Maintain providedDependencies of domains step without change

niranjan-uma-shankar committed:
Exclude domain only flow since site is not created

claudiucelfilip reviewed (approved):
Went through all scenarios and works as described; going back keeps the initial site and only updates the Order list on checkout.
Initially, I had my public-api sandbox, which caused a problem where I couldn't re-add the same domain if I removed it from the Order list. Fixed by un-sanboxing.

aneeshd16 commented:
>App Entrypoints (~35058 bytes added 📈 [gzipped])
>
>Details
>name        parsed_size            gzip_size
>entry-main    +132605 B  (+10.0%)   +35058 B  (+10.6%)

Hi @niranjan-uma-shankar, if I am reading it correctly, it says that around 132KB has been added to the main entry point which would cause longer load times. Is there anything that we can do to avoid this?

southp commented on client/my-sites/checkout/controller.jsx:
```diff
@@ -70,6 +71,19 @@ export function checkout( context, next ) {
 		'no-user' === context.query.cart;
 
+	const searchParams = new URLSearchParams( window.location.search );
+	const isSignupCheckout = searchParams.get( 'signup' ) === '1';
```
Could you help me point out where this is set in the first place? I'd like to confirm how confident we can rely on this.

niranjan-uma-shankar replied:
It is being set in [this line](https://github.com/Automattic/wp-calypso/blob/master/client/signup/config/flows.js#L26).

southp commented on client/my-sites/checkout/controller.jsx:
```diff
@@ -70,6 +71,19 @@ export function checkout( context, next ) {
+			const signupDestinationCookieExists = retrieveSignupDestination();
+
+			if ( signupDestinationCookieExists ) {
+				sessionStorage.setItem( 'signupCheckoutPageUnloaded', true ); // eslint-disable-line no-undef
```
Just like what you've done with the cookie, I'd suggest to abstract the access to this session storage flag as a pair of getter/setter in utils.js, so we can both name them as meaningful functions and keep them in one place for easier maintenance.

niranjan-uma-shankar replied:
Done - I've moved sessionStorage getters and setters to their own utils file.

southp commented on client/signup/main.jsx:
```diff
@@ -161,6 +165,23 @@ class Signup extends React.Component {
 		}
 
+		const searchParams = new URLSearchParams( window.location.search );
+		const isAddNewSiteFlow = searchParams.has( 'ref' );
```
I am not sure I understand this. Could you help me understand why we can rely on this?

niranjan-uma-shankar replied:
Clicking on "Add new site" from Calypso will take to the signup flow with a `ref` query parameter([code ref](https://github.com/Automattic/wp-calypso/blob/a2db298e5e1f43fa039fa1ec6a8667bb02a64099/client/my-sites/sidebar/index.jsx#L863)). If the ref exists, then the user clicked on "Add new site", and in this case we will delete the signup destination cookie since the user is already inside Calypso.

southp commented:
@niranjan-uma-shankar I have run through all the test scenarios and a few other combinations, and everything works perfectly. However, I'd need your help clarifying a few questions in the inlined comments so I can review more properly. Thanks :)

niranjan-uma-shankar commented:
>  if I am reading it correctly, it says that around 132KB has been added to the main entry point which would cause longer load times. Is there anything that we can do to avoid this?

@aneeshd16 Thanks for spotting that! I rebased today and the number has halved to ~60KB. Is that still a bit high for a change of this nature? I don't know, but I'm also not calling any packages.

aneeshd16 commented:
> @aneeshd16 Thanks for spotting that! I rebased today and the number has halved to ~60KB. Is that still a bit high for a change of this nature?

We did get a call-out for this when the size was 18KB, so this maybe an issue.
https://github.com/Automattic/wp-calypso/pull/44223#issuecomment-674829532
https://github.com/Automattic/wp-calypso/pull/44644#issuecomment-674839076

>  I don't know, but I'm also not calling any packages.

It is probably because additional imports are causing the bundles to be built in a different way - adding size to the main bundle while reducing it from other bundles. I am not sure specifically which import might cause this, but it could be this one:
https://github.com/Automattic/wp-calypso/pull/45040/files#diff-8302628106a075feef9e1f2c79429ff7R43

niranjan-uma-shankar committed:
Refactor to async load the checkout masterbar, this should help reduce the bundle size

niranjan-uma-shankar committed:
Test commit for moving cookie functions to its own utils section

niranjan-uma-shankar committed:
Test commit to know entry point size

niranjan-uma-shankar committed:
Move cookie handling utils to their own file and change dependencies

niranjan-uma-shankar committed:
Move cookie handling and sessionStorage utils to their own file and change dependencies

niranjan-uma-shankar commented:
> We did get a call-out for this when the size was 18KB, so this maybe an issue.

@aneeshd16 I refactored the code and moved the checkout masterbar to be async loaded, and moved the utils to another file. Looks like this has greatly helped reduce the size:

<img width="452" alt="Screenshot 2020-09-22 at 7 18 06 PM" src="https://user-images.githubusercontent.com/1269602/93890880-6daf6b80-fd08-11ea-93df-1599d01feef0.png">

niranjan-uma-shankar commented:
@Automattic/shilling I've added y'all as reviewers since this PR makes the masterbar in checkout page to be async loaded. This really helped reduce bundle size([check comment](https://github.com/Automattic/wp-calypso/pull/45040#issuecomment-696734223)). If you can confirm no issues with the checkout masterbar related confirm, that would be awesome. 

Besides wpcom, I verified the masterbar renders well in Jetpack too:

![signup](https://user-images.githubusercontent.com/1269602/93892087-c2071b00-fd09-11ea-90d3-fb1f265277b0.gif)

niranjan-uma-shankar committed:
Store siteSlug in session storage instead of creating a new cookie

niranjan-uma-shankar commented:
I pushed a cleanup in [12be1bc](https://github.com/Automattic/wp-calypso/pull/45040/commits/12be1bc5c8254601c43ad652d3a0d8e8622c350e) since I realized that I don't need to create yet another cookie when I could just use session storage.

niranjan-uma-shankar committed:
Check previousPath is not undefined since we are checking value using includes

sirbrillig reviewed (approved):
> this PR makes the masterbar in checkout page to be async loaded. This really helped reduce bundle size(check comment). If you can confirm no issues with the checkout masterbar related confirm, that would be awesome.

Thank you for asking! I tried it out, and while I didn't run through the full testing instructions, I don't see any issues with the masterbar in Jetpack or WPCOM checkout.

There will be a small conflict fix needed when https://github.com/Automattic/wp-calypso/pull/45558 lands because I moved `useRedirectIfCartEmpty` to its own file (sorry!) but ultimately that should help that hook be easier to work with.

aneeshd16 commented on client/signup/main.jsx:
```diff
@@ -161,6 +167,23 @@ class Signup extends React.Component {
+			clearSignupDestinationCookie();
+		}
+
+		const wasSignupCheckoutUnnloaded = getFromSessionStorage( 'signupCheckoutPageUnloaded' );
```
Do you think it will be better to have functions explicitly get/set these values rather than relying on strings? For example - `get(set)SignupCheckoutPageUnloaded()`, `get(set)SignupCompleteSlug()`. The wrapper functions otherwise do not extend `sessionStorage` functionality.

southp replied:
I'd like to second what @aneeshd16 proposed here, since it was actually what I meant when I left [this comment](https://github.com/Automattic/wp-calypso/pull/45040#discussion_r485482467) :)

Borrowing the terms from mathematics, the primary purpose of abstracting them into specific getter/setter here is to restrain the domains and the ranges of these functions so that they are unlikely to be misused. 

Also, I think `wasSignupCheckoutUnnloaded` has an redundant n here? i.e. Unnloaded -> Unloaded?

niranjan-uma-shankar replied:
Feedback addressed.

aneeshd16 reviewed (commented):
Tested these scenarios, looks good to me! I have one minor nitpick in the comments :)

Also, thanks for making the masterbar to be async loaded, it looks like a lot of savings. 🚀

southp commented on client/signup/storageUtils.js:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Exernal dependencies
```
Typo: Exernal -> External

niranjan-uma-shankar replied:
Thanks for catching that. I copied this line from utils.js, and I didn't notice the typo. I now fixed it in both files.

niranjan-uma-shankar committed:
Fix merge conflict

niranjan-uma-shankar committed:
Make the getters and setters specific

niranjan-uma-shankar committed:
Fix the getter to be a function instead of value

niranjan-uma-shankar committed:
Check if the user is re-entering the same flow, since we don't want to classify another flow as a managed site flow

niranjan-uma-shankar committed:
Fix typo

niranjan-uma-shankar committed:
Fix typo

niranjan-uma-shankar committed:
Refactor and move to functions, to improve readability; Add JSDoc
